parser.hh

```diff
@@ -199,9 +206,15 @@ class Input
 	 * so that the first request will trigger a fetch from the underlying
 	 * storage.
 	 */
-	Input() : buffer(0), buffer_start(1), buffer_end(0) {}
+	Input(std::string name)
```

Comment:
Doc comment should say what the parameter does.

Also, is there a reason for making this a copy, rather than reference in the parameter?  I don't think C++ does copy elision for initialisers, so this should probably either be a reference parameter (in which case the initialisation will copy the data), or use the move constructor for the field (i.e. `use_name(std::move(name))`).